Edit/editorial 2020 07 14 #5

lofstrandbjorn opened the pull request:
Please review Module + documentation.

ChristianPick commented:
General:
* You proposed to name all references ...Ref instead of ...Id. From my side this would be OK. Should we change the naming?

CommunicationNode:
* Semantics of EntityId: Why should this reference a platform or lifeform only? What about EnvironmentalObjects? E.g. a command post etc. could be represented by a container.
* NetworkDevices.PhysicalNetworkName type should be Text64
* NetworkDevices.TX/RX: Variants shoult be reduced to NoNetwork / Generic as this was done in the PhysicalNetwork too.

lofstrandbjorn commented:
Somehow I got an email with comments from the review of the PR but I can not find them here on GitHub.

General:
You proposed to name all references ...Ref instead of ...Id. From my side this would be OK. Should we change the naming?

CommunicationNode:

Semantics of EntityId: Why should this reference a platform or lifeform only? What about EnvironmentalObjects? E.g. a command post etc. could be represented by a container.
NetworkDevices.PhysicalNetworkName type should be Text64
NetworkDevices.TX/RX: Variants shoult be reduced to NoNetwork / Generic as this was done in the PhysicalNetwork too.

lofstrandbjorn commented:
> General:
> You proposed to name all references ...Ref instead of ...Id. From my side this would be OK. Should we change the naming?

I changed my mind and added a clarification in the semantics if it is an identifier or a reference. Ok as it is now?

lofstrandbjorn commented:
> CommunicationNode:
> 
> Semantics of EntityId: Why should this reference a platform or lifeform only? What about EnvironmentalObjects? E.g. a command post etc. could be represented by a container.
> NetworkDevices.PhysicalNetworkName type should be Text64
> NetworkDevices.TX/RX: Variants shoult be reduced to NoNetwork / Generic as this was done in the PhysicalNetwork too.

Will fix and update. Will add CulturalFeature to the list.

lofstrandbjorn commented:
> General:
> 
> * You proposed to rename ...Id to ...Ref for all references. For me this would be OK.
> 
> CommunicationNode:
> 
> * Why should EntityId reference a platform or lifeform only? Could be a cultural feature, too. Or a point object (as soon as the object model of NETN covers this, too).
> * In NetworkDevices: The PhysicalNetworkName should be of type Text64
> * In NetworkDevices: The variants of TX/RX could be reduced to NoNetwork / Generic as the other variants are currently not supported by the PhysicalNetwork.

Interesting. I got the email with the comment 5 min before is was posted on GitHub.

lofstrandbjorn commented:
> * In NetworkDevices: The variants of TX/RX could be reduced to NoNetwork / Generic as the other variants are currently not supported by the PhysicalNetwork.

The Generic RX variant is the same as NoNetwork. Ok?

ChristianPick commented:
> > * In NetworkDevices: The variants of TX/RX could be reduced to NoNetwork / Generic as the other variants are currently not supported by the PhysicalNetwork.
> 
> The Generic RX variant is the same as NoNetwork. Ok?

Maybe. ;-)
I haven't thought in detail what should be in the TX/RX or generic physicalnetwork structure.

ChristianPick commented:
> > General:
> > 
> > * You proposed to rename ...Id to ...Ref for all references. For me this would be OK.
> > 
> > CommunicationNode:
> > 
> > * Why should EntityId reference a platform or lifeform only? Could be a cultural feature, too. Or a point object (as soon as the object model of NETN covers this, too).
> > * In NetworkDevices: The PhysicalNetworkName should be of type Text64
> > * In NetworkDevices: The variants of TX/RX could be reduced to NoNetwork / Generic as the other variants are currently not supported by the PhysicalNetwork.
> 
> Interesting. I got the email with the comment 5 min before is was posted on GitHub. See comments above.

that's because I deleted the original post and placed it as a comment of the review. Less magic than one could think...

ThomasBrannstrom commented on NETN-COM.md:
```diff
@@ -1 +1,127 @@
+|Location|**Optional.** Specifies the location of the CommunicationNode in case the entity referenced by EntityId is not registered in the federation. If the entity referenced by EntityId exists in the federation, the location of the communication node is derived from that entity and the value of the Location attribute shall be ignored.|
+|RequestedConnections|**Required.** Possible (requested) connections for the communication node. |
+|NetworkDevices|**Required.** Available network devices defining the association of a communication network (connection layer) with a physical network (link layer). Each network device can be associated with several communication networks but only to one physical network. Each network device also describes the transmitter and receiver capabilities.|
+|IncomingConnections|**Optional.** Description of all incoming connections to the receiving entity.|
```
IncomingConnections and Recievers have different datastructure and stored in a different way.
IncomingConnections is stored in the CommunicationNode object and has the CommunicationNetworkName baked into its datastructure, while Recievers has its own Connection-object with CommunicationNetworkName in the Connection-object.
Why make it work different for those two sets of information, they are basically the same just mirrored?

ChristianPick replied:
What do you mean with "Receivers"? Each Connection object describes the outgoing connection of a sending entity (ConnectionNode) within the domain of a single communication network. I'm not sure what's your point.

ThomasBrannstrom replied:
On the Connection object there is a Receivers attribute represented as a ConnectionReceiverArray.
On the CommunicationNode there is a IncommingConnections attribute represented as an IncomingConnectionsArray.
My thought was that those two where "mirrors" of each others.
The Connection-object's Receivers attribute describes which nodes he can send data to.
The CommunicationNode-object's IncommingConnections attribute describes which nodes he can receive data from.
If that's the case, then they are "mirrors" of each other and maybe should be represented in similar way in the object model?

ChristianPick replied:
The design was driven by the idea that the sender specifies who he wants to talk to. A requested connection should lead to a Connection object (as soon as possible). To allow load balancing on the calculation of the outgoing connections objects instead of an array-attribute is used (otherwise we'll have some trouble with the ownership management of the attribute). Based on the Connection objects the information could be (optionally) aggregated into the IncomingConnectios attribute. Therefore this is a mirror of the information. As doing this type of aggregation is far less time consuming than the connection calculation it is assumed to be not critical and could be done by the receiver itself or a dedicated federate. Nevertheless the incoming connections attribute is optional. In my first proposal the IncomingConnections were an object like the outgoing connection. At the end the incoming connections attribute is for convenience as it is much easier to get the data needed by the receiver.

ThomasBrannstrom replied:
Lets keep it like it is, it will work for the other case we had as well.

lofstrandbjorn replied:
Then I consider this comment resolved.

ThomasBrannstrom commented on NETN-COM.md:
```diff
@@ -1 +1,127 @@
+## CommunicationNode
+A CommunicationNode is the representation of the interface of a simulated entity to logical communication networks. The location of the CommunicationNode is derived from the referenced entity or specified explicitly (if the referenced entity is not registered in the federation). 
+
+Connections to communication networks are only established if connectivity exists based on the status of available network devices and the simulation of links. Each potential connection is described in terms of requested connections  (the equivalent to the information provided by MSDL). 
```
“Connections to communication networks are only established if connectivity exists based on the status of available network devices and the simulation of links.”
--> Need to use PhysicalNetwork and need to have Links to establish a connection? Or are those objects optional?

ChristianPick replied:
As the application layer should never use the PhysicalNetwork or LinkStates objects they are optional. They are needed if you split the calculation of the link quality and the simulation of the overall connection. If both is done within one simulation the PhysicalNetwork and LinkStates objects are not needed as there is noone using them.

ThomasBrannstrom replied:
Thats good, I like that design.
Maybe the text that you wrote here should be included somehow in the documentation?
It doesn't say anywhere that those two objects are optional (I think).

lofstrandbjorn replied:
Thomas/Christian can you propose a text for this to be included in the .md!

ChristianPick replied:
A CommunicationNode is the representation of the interface of a simulated entity to logical communication networks. The location of the CommunicationNode is derived from the referenced entity or specified explicitly (if the referenced entity is not registered in the federation). 

Each potential connection is described in terms of requested connections  (the equivalent to the information provided by MSDL). The availability of a connection and its receivers are described by Connection objects. It is up to a communication simulation to create Connection objects for the requested connections based on the network devices of the CommunicationNode, the physical network and the link quality between network devices. The CommunicationNodes' federate has not to subscribe to PhysicalNetwork or LinkStates objects besides it has the role of the communication simulation and the use of these objects is mandatory by the federation agreement.

lofstrandbjorn replied:
I propose the following text based on your input:

"A `CommunicationNode` is the representation of the interface of a simulated entity to logical communication networks. The location of the `CommunicationNode` is derived from the referenced entity or specified explicitly if the referenced entity is not registered in the federation.

Each potential connection is described in terms of requested connections (the equivalent to the information provided by MSDL). `Connection` objects represent the availability of a connection and its receivers. It is up to a communication simulation to create `Connection` objects for the requested connections based on the network devices of the `CommunicationNode`, the physical network and the link quality between network devices."

The last sentence in your proposed text I do not understand.

"The CommunicationNodes' federate has not to subscribe to PhysicalNetwork or LinkStates objects besides it has the role of the communication simulation and the use of these objects is mandatory by the federation agreement."

I think you mean something like "Depending on the federation design and agreements, the explicit representation of the physical layer objects, `PhysicalNetwork` and `LinkStates`, is optional."

ChristianPick replied:
Yes

lofstrandbjorn replied:
Ok I will make the following change and then consider this comment resolved.

"A `CommunicationNode` is the representation of the interface of a simulated entity to logical communication networks. The location of the `CommunicationNode` is derived from the referenced entity or specified explicitly if the referenced entity is not registered in the federation.

Each potential connection is described in terms of requested connections (the equivalent to the information provided by MSDL). `Connection` objects represent the availability of a connection and its receivers. It is up to a communication simulation to create `Connection` objects for the requested connections based on the network devices of the `CommunicationNode`, the physical network and the link quality between network devices. Depending on the federation design and agreements, the explicit representation of the physical layer objects, `PhysicalNetwork` and `LinkStates`, is optional."

ThomasBrannstrom commented on NETN-COM.md:
```diff
@@ -1 +1,127 @@
+|UniqueName|**Required.** Unique physical network name. Uniqueness in the context of physical networks.|
+|Description|**Required.** Characteristics of the physical network.|
+
+## LinkStates
```
LinkStatusArray FOM semantics: “The status of a physical network link.”
--> **Should probably be:** “The status of physical network link**s**.”
(since LinkStatusArray is an array of links between transmitter and receivers?

lofstrandbjorn replied:
Ok I will update.

ThomasBrannstrom commented on NETN-COM.md:
```diff
@@ -1 +1,127 @@
+# NETN-COM
```
Would be nice to get an updated Overview picture, the one with colored boxes.

ChristianPick replied:
@lofstrandbjorn: The picture is not included in the current branch. If you include it I can update it...

lofstrandbjorn replied:
I pushed update included png in the image subfolder and added an img link in .md file.

lofstrandbjorn replied:
Unless you are working on it now, I suggest to open a separate issue for updating the picture and resolve this comment for now in order to proceed to merge the PR.

ChristianPick replied:
I have trouble syncing my Fork - would it be possible to work on the branch directly? I have no access rights...

ChristianPick replied:
...or just update the picture by this one:

![Concepts](https://user-images.githubusercontent.com/67922322/88366523-a1622980-cd89-11ea-888d-186375d14caa.png)

lofstrandbjorn replied:
Only members of MSG-163 have the write access to repos. A formal NATO MS3 thing. You are welcome to join :-) In the mean time made the changes and will now resolve this issue.

ChristianPick commented on NETN-COM.md:
```diff
@@ -1 +1,127 @@
+
+The model does not require all levels and networks to be represented in the federation. Which objects are needed depends on the federation design and allocation of modelling responsibilities. E.g. at the application layer, only the connection information needs to be published.
+
+The application layer should only use the `CommunicationNode` and `Connection` objects to determine if data can be sent or received. I.e. data should only be sent if a valid instance of `Connection` for transmitting data is available, and receivers should reject data from a sender if it is not included in `IncomingConnections`.
```
I havn't reviewed the .md file before. The description is indeed misleading. As IncomingConnections are optional the behavior should be as follows:

"The application layer should only use 'CommunicationNode' and 'Connection' objects to determine if data can be sent or received. If no Connection object is present or no receiver is set set the sender could drop the data. The receiver has to reject incoming messages if he is not inlcuded in the receivers' list of the corresponding Connection object. In case IncommingConnections are calculated for the receivers' CommunicationNode this data could be used instead of the data present in the Connection object."

In theory a design without the Connection objects is possible if you have a single one-fits-all communication simulation directly calculating all IncomingConnection attributes for all CommunicationNodes. This is the way the previous design was. We could discuss whether this option should still be available - in such a case it would be up to the federation agreement to define which way to use.

ThomasBrannstrom replied:
Yeah, we had that up for discussion before as well, about running it in two different modes.
It's not **needed** for our case I think, but could be a bit easier.

lofstrandbjorn replied:
I propose the following text:

"The application layer should only use the `CommunicationNode` and `Connection` objects to determine if data can be sent or received. If no `Connection` object exists or if the `Receivers` attribute is empty, the sender can drop the data entirely. A receiver should drop any incoming messages if not included in the `Receivers` attribute of the corresponding `Connection` object. If the `IncommingConnections` attribute of a `CommunicationNode` is calculated for a receiver, it can be used to determine if incoming data should be dropped or not."

Ok?

ChristianPick replied:
yes

lofstrandbjorn replied:
Ok. implemented. Comment resolved.

lofstrandbjorn commented on NETN-COM.md:
```diff
@@ -1 +1,127 @@
+# NETN-COM
+NATO Education and Training Network (NETN) Communication (COM) Module. 
+        
+This module is a specification of how to represent Networks and related connectivity data to be shared among participants in a federated distributed simulation. The specification is based on IEEE 1516 High Level Architecture (HLA) Object Model Template (OMT) and primarily intended to support interoperability in a federated simulation (federation) based on HLA. An HLA OMT based Federation Object M ...
```
Changed in doc & readme files.

lofstrandbjorn commented on NETN-COM.md:
```diff
@@ -1 +1,127 @@
+        
 
+## Purpose
+The purpose of the NETN COM module is to provide a standard way to exchange data related to Communication Networks and Physical Networks. The main objective is to provide a generic model that represents the status of connections in a communication network and links in a physical network.
```
Added

lofstrandbjorn commented:
All changes implemented and committed to the PR branch. Ready for merge?
